integrate/romberg.go

```diff
@@ -0,0 +1,60 @@
+// Copyright ©2019 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package integrate
+
+import (
+	"math"
+	"math/bits"
+)
+
+// Romberg estimates the integral of the function f
+// \int_a^b f(x)dx
+// The function f must be sampled at equal intervals given by dx.
+// Romberg's method approximates the integral of the function by
+// applying Richardson extrapolation.
+// See https://en.wikipedia.org/wiki/Romberg%27s_method for a thorough description
+// of the algorithm.
+//
+// The length of the length of the f array must be 2**k + 1, where k is a positive integer
+// The method returns the value of the integral
+func Romberg(f []float64, dx float64) float64 {
+	if len(f) < 3 {
+		panic("integral: invalid slice length: must be 2^k + 1")
+	}
+	ninterv := len(f) - 1
+	k := bits.Len(uint(ninterv - 1))
+
+	n := 1 << uint(k)
+	if n != ninterv {
+		panic("integral: invalid slice length: must be 2^k + 1")
+	}
+
+	work := make([]float64, 2*k)
+	prev := work[:k]
+	curr := work[k:]
+
+	h := dx * float64(len(f))
```

Comment:
Is `h` supposed to be equal to `b - a`? If that's the case, this should have `len(f)-1`. It would be nice to have a test that fails due to this; it will need a test with few samples to make a difference of one `dx` significant.

Reply:
With `n = 1<<8 + 1` this bug demonstrates itself.

Reply:
There is now a test with only 5 points. However, it tests a polynomial function as then we now the answer should be exact.